File: EntativaBackend/services/notification-service/src/main/scala/com/vignette/notification/service/EmailService.scala
package com.socialink.notification.service

import java.util.Properties
import javax.mail._
import javax.mail.internet._
import scala.util.{Try, Success, Failure}

/**
 * Email notification service
 * Sends email notifications with HTML templates
 */
class EmailService(
  smtpHost: String,
  smtpPort: Int,
  username: String,
  password: String,
  fromAddress: String
) {
  
  private val props = new Properties()
  props.put("mail.smtp.host", smtpHost)
  props.put("mail.smtp.port", smtpPort.toString)
  props.put("mail.smtp.auth", "true")
  props.put("mail.smtp.starttls.enable", "true")
  
  private val session = Session.getInstance(props, new Authenticator {
    override def getPasswordAuthentication: PasswordAuthentication = {
      new PasswordAuthentication(username, password)
    }
  })
  
  /**
   * Send email
   */
  def sendEmail(to: String, subject: String, body: String, isHTML: Boolean = true): Either[String, Unit] = {
    Try {
      val message = new MimeMessage(session)
      message.setFrom(new InternetAddress(fromAddress))
      message.setRecipients(Message.RecipientType.TO, to)
      message.setSubject(subject)
      
      if (isHTML) {
        message.setContent(body, "text/html; charset=utf-8")
      } else {
        message.setText(body)
      }
      
      Transport.send(message)
      
    } match {
      case Success(_) => Right(())
      case Failure(ex) => Left(s"Email error: ${ex.getMessage}")
    }
  }
  
  /**
   * Send templated email
   */
  def sendTemplatedEmail(
    to: String,
    templateName: String,
    data: Map[String, String],
    subject: String
  ): Either[String, Unit] = {
    val body = renderTemplate(templateName, data)
    sendEmail(to, subject, body, isHTML = true)
  }
  
  /**
   * Send batch emails
   */
  def sendBatch(recipients: Seq[String], subject: String, body: String): Either[String, Int] = {
    var successCount = 0
    var errors = List.empty[String]
    
    recipients.foreach { recipient =>
      sendEmail(recipient, subject, body) match {
        case Right(_) => successCount += 1
        case Left(error) => errors = error :: errors
      }
    }
    
    if (errors.isEmpty) {
      Right(successCount)
    } else {
      Left(s"$successCount/${recipients.size} sent. Errors: ${errors.take(5).mkString(", ")}")
    }
  }
  
  /**
   * Render email template
   */
  private def renderTemplate(templateName: String, data: Map[String, String]): String = {
    templateName match {
      case "notification" =>
        s"""
          <!DOCTYPE html>
          <html>
          <head>
            <style>
              body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; margin: 0; padding: 0; background: #f5f5f5; }
              .container { max-width: 600px; margin: 40px auto; background: white; border-radius: 12px; overflow: hidden; box-shadow: 0 4px 6px rgba(0,0,0,0.1); }
              .header { background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; }
              .content { padding: 30px; }
              .actor { display: flex; align-items: center; margin-bottom: 20px; }
              .avatar { width: 48px; height: 48px; border-radius: 50%; margin-right: 12px; background: #667eea; }
              .message { font-size: 18px; color: #333; line-height: 1.6; }
              .button { display: inline-block; background: #667eea; color: white; padding: 14px 28px; text-decoration: none; border-radius: 8px; margin-top: 20px; font-weight: 600; }
              .footer { background: #f9f9f9; padding: 20px; text-align: center; color: #666; font-size: 14px; }
            </style>
          </head>
          <body>
            <div class="container">
              <div class="header">
                <h1>📱 ${data.getOrElse("title", "New Notification")}</h1>
              </div>
              <div class="content">
                ${data.get("actorUsername").map(u => s"""
                <div class="actor">
                  <div class="avatar"></div>
                  <div><strong>@$u</strong></div>
                </div>
                """).getOrElse("")}
                <div class="message">
                  ${data.getOrElse("message", "You have a new notification")}
                </div>
                ${data.get("deepLink").map(link => s"""
                <a href="$link" class="button">View Now</a>
                """).getOrElse("")}
              </div>
              <div class="footer">
                <p><strong>Socialink</strong> - Share Your Moments</p>
                <p><small>Manage notification settings in your app</small></p>
              </div>
            </div>
          </body>
          </html>
        """
      
      case "weekly_digest" =>
        s"""
          <!DOCTYPE html>
          <html>
          <head>
            <style>
              body { font-family: Arial, sans-serif; }
              .container { max-width: 600px; margin: 0 auto; }
              .stat { background: #f0f0f0; padding: 15px; margin: 10px 0; border-radius: 8px; }
            </style>
          </head>
          <body>
            <div class="container">
              <h2>Your Weekly Highlights</h2>
              <div class="stat">
                <h3>${data.getOrElse("likes", "0")} Likes</h3>
              </div>
              <div class="stat">
                <h3>${data.getOrElse("comments", "0")} Comments</h3>
              </div>
              <div class="stat">
                <h3>${data.getOrElse("followers", "0")} New Followers</h3>
              </div>
            </div>
          </body>
          </html>
        """
      
      case _ =>
        s"<html><body><p>${data.getOrElse("message", "")}</p></body></html>"
    }
  }
}

File: VignetteBackend/services/user-service/internal/service/email_service.go
package service

import (
	"bytes"
	"fmt"
	"html/template"
	"log"
	"net/smtp"
	"os"
)

// EmailService handles sending emails
type EmailService struct {
	smtpHost     string
	smtpPort     string
	smtpUsername string
	smtpPassword string
	fromEmail    string
	fromName     string
}

// NewEmailService creates a new email service instance
func NewEmailService() *EmailService {
	return &EmailService{
		smtpHost:     getEnv("SMTP_HOST", "smtp.gmail.com"),
		smtpPort:     getEnv("SMTP_PORT", "587"),
		smtpUsername: getEnv("SMTP_USERNAME", ""),
		smtpPassword: getEnv("SMTP_PASSWORD", ""),
		fromEmail:    getEnv("FROM_EMAIL", "noreply@entativa.com"),
		fromName:     getEnv("FROM_NAME", "Entativa"),
	}
}

// SendPasswordResetEmail sends a password reset email
func (s *EmailService) SendPasswordResetEmail(toEmail, firstName, resetLink string) error {
	subject := "Reset Your Entativa Password"
	
	// HTML template for the email
	htmlTemplate := `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: linear-gradient(135deg, #007CFC 0%, #6F3EFB 50%, #FC30E1 100%); color: white; padding: 30px; text-align: center; border-radius: 8px 8px 0 0; }
        .content { background: white; padding: 30px; border: 1px solid #e4e6eb; border-top: none; border-radius: 0 0 8px 8px; }
        .button { display: inline-block; background: #007CFC; color: white; padding: 14px 32px; text-decoration: none; border-radius: 8px; margin: 20px 0; font-weight: 600; }
        .footer { text-align: center; color: #65676b; font-size: 12px; margin-top: 20px; }
        .warning { background: #fff3cd; border: 1px solid #ffc107; padding: 15px; border-radius: 6px; margin: 20px 0; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1 style="margin: 0; font-size: 36px; font-style: italic;">entativa</h1>
        </div>
        <div class="content">
            <h2>Hi {{.FirstName}},</h2>
            <p>We received a request to reset your Entativa password. Click the button below to create a new password:</p>
            <p style="text-align: center;">
                <a href="{{.ResetLink}}" class="button">Reset Password</a>
            </p>
            <div class="warning">
                <strong>⚠️ Important:</strong> This link will expire in 1 hour and can only be used once.
            </div>
            <p>If you didn't request this, you can safely ignore this email. Your password won't change unless you click the link above.</p>
            <p>For security reasons, we never send your password via email.</p>
            <hr style="border: none; border-top: 1px solid #e4e6eb; margin: 30px 0;">
            <p style="color: #65676b; font-size: 14px;">
                If the button doesn't work, copy and paste this link into your browser:<br>
                <a href="{{.ResetLink}}" style="color: #007CFC; word-break: break-all;">{{.ResetLink}}</a>
            </p>
        </div>
        <div class="footer">
            <p>© 2025 Entativa. All rights reserved.</p>
            <p>This is an automated email. Please do not reply.</p>
        </div>
    </div>
</body>
</html>
`
	
	// Parse template
	tmpl, err := template.New("passwordReset").Parse(htmlTemplate)
	if err != nil {
		return fmt.Errorf("failed to parse email template: %w", err)
	}
	
	// Execute template
	var body bytes.Buffer
	data := struct {
		FirstName string
		ResetLink string
	}{
		FirstName: firstName,
		ResetLink: resetLink,
	}
	
	if err := tmpl.Execute(&body, data); err != nil {
		return fmt.Errorf("failed to execute email template: %w", err)
	}
	
	// Send email
	return s.sendEmail(toEmail, subject, body.String())
}

// SendWelcomeEmail sends a welcome email to new users
func (s *EmailService) SendWelcomeEmail(toEmail, firstName string) error {
	subject := "Welcome to Entativa!"
	
	htmlTemplate := `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <style>
        body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; }
        .container { max-width: 600px; margin: 0 auto; padding: 20px; }
        .header { background: linear-gradient(135deg, #007CFC 0%, #6F3EFB 50%, #FC30E1 100%); color: white; padding: 40px; text-align: center; border-radius: 8px 8px 0 0; }
        .content { background: white; padding: 30px; border: 1px solid #e4e6eb; border-top: none; border-radius: 0 0 8px 8px; }
        .button { display: inline-block; background: #007CFC; color: white; padding: 14px 32px; text-decoration: none; border-radius: 8px; margin: 20px 0; font-weight: 600; }
        .footer { text-align: center; color: #65676b; font-size: 12px; margin-top: 20px; }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1 style="margin: 0; font-size: 48px; font-style: italic;">entativa</h1>
            <p style="font-size: 18px; margin-top: 10px;">Connect with friends and the world</p>
        </div>
        <div class="content">
            <h2>Welcome, {{.FirstName}}! 🎉</h2>
            <p>Your Entativa account has been successfully created. We're excited to have you join our community!</p>
            <p>Here's what you can do next:</p>
            <ul>
                <li>Complete your profile</li>
                <li>Connect with friends</li>
                <li>Share your first post</li>
                <li>Explore groups and communities</li>
            </ul>
            <p style="text-align: center;">
                <a href="https://app.entativa.com" class="button">Get Started</a>
            </p>
        </div>
        <div class="footer">
            <p>© 2025 Entativa. All rights reserved.</p>
        </div>
    </div>
</body>
</html>
`
	
	tmpl, err := template.New("welcome").Parse(htmlTemplate)
	if err != nil {
		return err
	}
	
	var body bytes.Buffer
	data := struct {
		FirstName string
	}{
		FirstName: firstName,
	}
	
	if err := tmpl.Execute(&body, data); err != nil {
		return err
	}
	
	return s.sendEmail(toEmail, subject, body.String())
}

// sendEmail sends an email using SMTP
func (s *EmailService) sendEmail(to, subject, htmlBody string) error {
	// In development, just log the email
	if s.smtpUsername == "" || s.smtpPassword == "" {
		log.Printf("📧 [DEV MODE] Email would be sent to: %s\nSubject: %s\n", to, subject)
		return nil
	}
	
	// Build email message
	from := fmt.Sprintf("%s <%s>", s.fromName, s.fromEmail)
	message := []byte(fmt.Sprintf("From: %s\r\n"+
		"To: %s\r\n"+
		"Subject: %s\r\n"+
		"MIME-Version: 1.0\r\n"+
		"Content-Type: text/html; charset=UTF-8\r\n"+
		"\r\n"+
		"%s\r\n", from, to, subject, htmlBody))
	
	// Set up authentication
	auth := smtp.PlainAuth("", s.smtpUsername, s.smtpPassword, s.smtpHost)
	
	// Send email
	addr := fmt.Sprintf("%s:%s", s.smtpHost, s.smtpPort)
	err := smtp.SendMail(addr, auth, s.fromEmail, []string{to}, message)
	if err != nil {
		return fmt.Errorf("failed to send email: %w", err)
	}
	
	log.Printf("✅ Email sent to: %s", to)
	return nil
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}
